### ISSUE TITLE
Serbian Latin/Cyrillic translations

### ISSUE THREAD

**Issue by DominikNovosel** [m-i-n-a-r/birday#134]
This is a continuation of the discussion at https://github.com/m-i-n-a-r/birday/pull/129. We're looking to support both writing systems used by Serbian, but we aren't quite sure how, due to the locale naming scheme.

**Comment by vatonbero** [comment]
Hi there,

I helped translating Flym app to Serbian (cyrl/latn) and had the chance to see naming scheme there. Reviewing its source code might help: https://github.com/FredJul/Flym/tree/master/app/src/main/res

**Comment by DominikNovosel** [comment]
Thanks! I've never seen pluses in locale names, though. Has something changed in the meantime? There's actually a surprising amount of debate on this: https://github.com/moment/moment/issues/3899.
Still, everything seems to either be based on ISO country and language codes or BCP 47. I'm still trying to track down the official Android documentation on the matter, a list of locales I found on Stack Overflow wasn't helpful: 
```sr_ [Serbian]
sr_ [Serbian (Cyrillic)]
sr_BA [Serbian (Cyrillic,Bosnia and Herzegovina)]
sr_ME [Serbian (Cyrillic,Montenegro)]
sr_RS [Serbian (Cyrillic,Serbia)]
sr_XK [Serbian (Cyrillic,Kosovo)]
sr_ [Serbian (Latin)]
sr_BA [Serbian (Latin,Bosnia and Herzegovina)]
sr_ME [Serbian (Latin,Montenegro)]
sr_RS [Serbian (Latin,Serbia)]
sr_XK [Serbian (Latin,Kosovo)]```

Does this perhaps mean that Android automatically recognizes the script? That'd be silly, because I don't think we can have two locale folders with the same name.
I think that @Th3Unkn0wn0n3 's initial suggestion might be our best bet: ```
According to localeplanet code for cyrillic is sr_Cyrl, and for latin sr_Latn
```

**Comment by DominikNovosel** [comment]
This is more like it: https://github.com/moment/moment/issues/3899#issuecomment-293442214
Evidently, I'm not the only one to find this helpful as I'm the third person to mention this issue 🤣.
They make mention of BCP-47, here are the relevant documents:
https://www.rfc-editor.org/info/bcp47
https://www.rfc-editor.org/rfc/rfc5646.html

**Comment by DominikNovosel** [comment]
Side note, looking through my phone's settings, I only found the Cyrillic variant. Just thought that was interesting and a little strange.
![Screenshot_2022-05-28-20-15-13-981_com.android.settings.jpg](https://user-images.githubusercontent.com/66269443/170838097-4d9b04e3-71a8-4e1b-b0be-0b7af547b30c.jpg)

**Comment by vatonbero** [comment]
@DominikNovosel maybe we can start by creating 2 directories:
- values-sr
- values-b+sr+Latn

After that we can check. On my Android 11, Motorola phone, there are Latin and Cyrillic variants for Serbian to choose.

I have no LineageOS phones right now (last one used was Lineage 17) but inspecting their github repo and one random package, naming convention is the same as in Flym: https://github.com/LineageOS/android_packages_apps_Settings/tree/lineage-19.1/res

**Comment by DominikNovosel** [comment]
Sounds good! Do you happen to know what the "b" represents? I'm still pretty busy, do you know how to do this yourself?

**Comment by vatonbero** [comment]
> Sounds good! Do you happen to know what the "b" represents?

Some of that is explained on following pages:
- https://developer.android.com/training/basics/supporting-devices/languages#CreateDirs
- https://developer.android.com/guide/topics/resources/providing-resources#AlternativeResources

> I'm still pretty busy, do you know how to do this yourself?

If you are asking can i create PR with changed dir paths and to convert cyrl to latn, that I can. If you are asking me to build app with those changes after PR approval, I have no sdk configured on my personal workstation.

**Comment by DominikNovosel** [comment]
I suggest making the changes in your fork of Birday, I'll clone it, compile it and send you the APK here. Sound good?

**Comment by vatonbero** [comment]
> I suggest making the changes in your fork of Birday, I'll clone it, compile it and send you the APK here. If everything works out, you can make the pull request. Sound good?

Sounds like a plan! I will inform you when I do that.

**Comment by DominikNovosel** [comment]
Great! You don't have to do it all at once, I can help with the transliteration part, you just let me know if I accidentally do something stupid, such as leave out the conjunction "da" and give a verb the suffix "i", like we do in Croatian 😅.

**Comment by DominikNovosel** [comment]
Oh, wait, so you're **actually** done? Already? Whoa. I'll be right back.

**Comment by DominikNovosel** [comment]
@vatonbero Here ya go, fresh from the oven 🍪. With fresh ingredients, too, I had a bunch of updates to wait on because I hadn't opened Android Studio in a while.
[app-debug.zip](https://github.com/m-i-n-a-r/birday/files/8791799/app-debug.zip)

**Comment by vatonbero** [comment]
> Great! You don't have to do it all at once, I can help with the transliteration part, you just let me know if I accidentally do something stupid, such as leave out the conjunction "da" and give a verb the suffix "i", like we do in Croatian sweat_smile. Besides, to start out, we just need some of the more prominent strings to see whether things are working.

To be frank, i just swapped Cyrillic chars with Latin ones, for the purpose of this test. I relied to existing translation. Usually translating to Serbian means to translate something to Cyrillic and then swap letter to Latin. I mean, i did that for at least 10 apps translated so far this way or another.



> @vatonbero Here ya go, fresh from the oven cookie. With fresh ingredients, too, I had a bunch of updates to wait on because I hadn't opened Android Studio in a while. [app-debug.zip](https://github.com/m-i-n-a-r/birday/files/8791799/app-debug.zip)

aaaand we have a winner:it works like a charm! Thank you for your swift action.


So to recap: it works now for both cases of Serbian. Maybe somebody else can try this as well. From my experience, I would suggest to try on Samsung phones, OneUI enabled ones (so no S7 or something that old), because they have only one Serbian and that is Latin. That test would prove that strings are being used correctly.

**Comment by DominikNovosel** [comment]
> To be frank, i just swapped Cyrillic chars with Latin ones, for the purpose of this test.

I know, but it still seemed like it would've taken a bit longer than it did. Is there some kind of automated way to do this?

> aaaand we have a winner:it works like a charm! Thank you for your swift action.

Good to hear, glad I helped!

> I would suggest to try on Samsung phones, OneUI enabled ones (so no S7 or something that old)

Hmm... I _do_ have a Samsung Galaxy S21 FE to hand, I'll definitely try at some point and let you know.

**Comment by vatonbero** [comment]
> I know, but it still seemed like it would've taken a bit longer than it did. Is there some kind of automated way to do this?

Depends on the OS and available shells (well, not anymore with WSL). You can do it online or make `tr` on linux to do that work for ya and create substitution for you like:

```
tr 'абвгдђ' 'abvgdđ'
```

I am usually lazy so I convert it online. Better ways are to use something like https://linux.die.net/man/1/recode-sr-latin

**Comment by DominikNovosel** [comment]
Interesting, that will certainly help me next time I need to read a huge chunk of Cyrillic text.

**Comment by DominikNovosel** [comment]
Just tried Birday on the Samsung Galaxy S21 FE. Houston, we've got a problem.

![Screenshot_20220602-133051_Settings.jpg](https://user-images.githubusercontent.com/66269443/171620819-6cad4dbe-0887-4321-b52c-0adf428c2f5d.jpg)
![Screenshot_20220602-133144_Birday.jpg](https://user-images.githubusercontent.com/66269443/171620843-09398d51-957f-4260-ba24-94ed76a8079b.jpg)

**Comment by m-i-n-a-r** [comment]
Excuse my dumbness, but where's the problem? Wrong alphabet maybe?

**Comment by DominikNovosel** [comment]
Yes, we were testing whether the Latin Serbian strings would show properly if the phone's system language was set to Latin Serbian. Of note, parts of the OS itself were displayed both in Cyrillic and Latin, almost at random.

**Comment by Th3Unkn0wn0n3** [comment]
We learn both scripts in primary school, everyone is supposed to know then

**Comment by DominikNovosel** [comment]
Is it normal for Android to display the UI in mix and match scripts in Serbian?

**Comment by vatonbero** [comment]
No, it is not normal for system to show mixed content however most of translated applications are not really in the right place. I mostly use Cyrillic but VLC, k-9 and Feeder are reverting to English language in apps if you select Serbian Latin. Never really noticed this but it is quite sad. @DominikNovosel can you try VLC and K-9 and see what language is displayed on your Samsung device?

As for preference: I dont think that there is such thing since we use them both and that choice is and should be highly independent given on what is available. From UX perspective, it would not be good to create only one translation type because it can loose consistency, compared to the rest of the system. e might conclude something someone else tries to install VLC and k-9 for example, set OS lang to Serbian (latin) and then Serbian (cyrillic) and reply back with findings.

It could be a problem with Samsung android skin (OneUI)? It works on vanilla(ish) Android so I will test created program in the upcoming days in MIUI (A12 based) and LineageOS (19.1, A12 based again). I should have all available resources in the next 10 to 14 days, if somebody doesn't beat me to it.

**Comment by vatonbero** [comment]
Weird. Well, i completed my tests:
1. Miui 13.0.6.0 - has only Cyrillic Serbian and app translation showed correct values in Cyrillic
2. LineageOS 19.1 - has options to show Serbian Latin and Cyrillic, both shown correctly in the birday app.

So this is interesting. We could conclude the following:
1. On systems where we have only one Serbian language variant, system will always show Cyrillic
2. On other clean (near vanilla) Android implementations (LineageOS, Motorola Android, possibly Nokia given how tightly the are chained to vanilla) system will honour changes and show appropriate language for birday

That gives us a few possible solutions:
1. Leave everything as is. Add Cyr and Latn variants and wait for upstream to adopt naming references. I stumbled upon the fact that many Chinese variants are implemented into Android the same way Serbian is
2. Leave only Cyrillic - not really a good offer here but it is practical from translators point of view: translating from Cyrilic to Latin is one click/pipe because we literally substitute letters and not change anything in the order. Maybe someday Android translation efforts will be supported by these Android "skins". Kinda defeats the purpose but it is worth mentioning. Given that Cyrillic is official variant used in all government related affairs, I would leave Cyrillic in any case. Again, we use both variants (most of us daily) but it if we had to choose between the two...

**Comment by DominikNovosel** [comment]
It seems reasonable to leave in both scripts since they're *sometimes* displayed correctly. However, I'm not at all confident smartphone manufacturers and their custom OS-es will ever stop screwing over both the users and the developers.

**Comment by DominikNovosel** [comment]
You lucky duck! At least you have a device compatible with Lineage OS! 😜

**Comment by vatonbero** [comment]
> So... are we done here? Should we close the issue? Seems to me we did everything that could be done.

I would say so. The logic will be there and @Th3Unkn0wn0n3 completed translation so issue is resolved.

Now maybe we will contact Samsung this year about this if there is an easy way to report this.

**Comment by DominikNovosel** [comment]
> Now maybe we will contact Samsung this year about this if there is an easy way to report this.

I don't think there is any easy way given Samsung's size, but this might help: https://stackoverflow.com/questions/4928267/where-to-report-device-specific-bugs-to-samsung

If you manage to resolve this, feel free to contact me again through this issue and I'll be happy to test it out. Good luck, and it's been great fun working with you!

**Comment by DominikNovosel** [comment]
Hey @vatonbero , I just noticed you didn't submit a pull request to include the Latin transliteration in the upstream repository, any particular reason?

**Comment by m-i-n-a-r** [comment]
That's why i wasn't aware of this contribution 😄 Submit a pull request if you want, i'll gladly credit you!
